src/launch/launcher.c

```diff
@@ -1155,6 +1155,55 @@ static int launcher_set_policy(Launcher *launcher, Policy *policy, uint32_t *sys
         return 0;
 }
 
+/**
+ * launcher_apparmor_check() - checks if AppArmor should be used
+ * @apparmor_mode:        bi-directional argument telling if AppArmor should be enabled
+ *
+ * This should be called with the requested apparmor mode. The
+ * function modifies the mode to either CONFIG_APPARMOR_DISABLED
+ * or CONFIG_APPARMOR_ENABLED based on platform support. If the
+ * function keeps the mode as CONFIG_APPARMOR_REQUIRED, AppArmor
+ * has been configured as mandatory but is not supported.
+ *
+ * Returns: 0 if check succeeded, or negative error code on failure.
+ */
+int launcher_apparmor_check(unsigned int *apparmor_mode) {
+        bool enabled, supported;
+        int r;
+
+        if (*apparmor_mode == CONFIG_APPARMOR_DISABLED)
+                return CONFIG_APPARMOR_DISABLED;
+
+        r = bus_apparmor_is_enabled(&enabled);
+        if (r)
+                return error_fold(r);
+
+        r = bus_apparmor_dbus_supported(&supported);
+        if (r)
+                return error_fold(r);
+
+        if (*apparmor_mode == CONFIG_APPARMOR_ENABLED) {
+                if (enabled && !supported) {
+                        fprintf(stderr, "Kernel is missing AppArmor DBus support.\n");
+                        *apparmor_mode = CONFIG_APPARMOR_DISABLED;
+                } else if (enabled) {
+                        *apparmor_mode = CONFIG_APPARMOR_ENABLED;
+                } else {
+                        *apparmor_mode = CONFIG_APPARMOR_DISABLED;
+                }
+        } else if (*apparmor_mode == CONFIG_APPARMOR_REQUIRED) {
+                if (!enabled || !supported) {
+                        fprintf(stderr, "AppArmor required, but not supported. Exiting.\n");
+                        *apparmor_mode = CONFIG_APPARMOR_REQUIRED;
```

Comment:
I expected to see a negative return value here, or perhaps an `exit(3)` or `_exit(2)` call.

Reply:
exit is handled outside of this function on line 1250

perhaps the fprintf on line 1196 should move between lines 1250 and 1251 ?